File: python/fibonacci_number.py
# no.505
class Solution:
    def __init__(self):
        self.cache = {}

    def fib(self, n: int) -> int:
        if n < 2:
            return n
        elif n in self.cache:
            return self.cache[n]
        else:
            ans = self.fib(n - 1) + self.fib(n - 2)
            self.cache[n] = ans
            return ans

File: C++/fibonacci_number.cpp
//no. 505
#include<iostream>
#include <map>
using namespace std;

class Solution {
    public:
    map<int, int> cache;
    Solution(){ 
        cache[0] = 0;
        cache[1] = 1;
    }
    int fib(int n) {
        map<int, int>:: iterator it;
        it = cache.find(n);
        if (it != cache.end()){
            return it->second;
        }
        int ans = fib(n-1) + fib(n-2);
        cache[n] = ans;
        return ans;
                        
    }
};

int main(){
    Solution sol;
    int k;
    cout << "enter question, k=";
    cin >> k;
    cout << "answer: " << sol.fib(k) << endl;
    return 0;
}
